src/EnergyPlus/SurfaceGeometry.cc

```diff
@@ -9300,6 +9300,65 @@ namespace SurfaceGeometry {
         }
     }
 
+    struct PopCoincidentVertexReturn
+    {
+        // bool popNeeded;
+        double perimeter;
+        int poppedVertexPos = -1; // This is a STL vector position, 0-indexed
+        int keptVertexPos = -1;
+    };
+
+    PopCoincidentVertexReturn checkPopCoincidentVertex(const Array1D<Vector> &vertices)
+    {
+        constexpr double tolerance = 0.01;
+
+        size_t const nSides = vertices.size();
+
+        // Vector of distance from this vertex to the next one
+        std::vector<Real64> distances(nSides);
+        size_t index = 0;
+        double min_distance = std::numeric_limits<Real64>::max();
+        double perimeter = 0.0;
+        for (auto it = vertices.begin(); it != vertices.end(); ++it) {
+            auto itnext = std::next(it);
+            if (itnext == std::end(vertices)) {
+                itnext = std::begin(vertices);
+            }
+            const auto dist = distance(*it, *itnext);
+            distances[index++] = dist;
+            min_distance = std::min(min_distance, dist);
+            perimeter += dist;
+        }
+        if (min_distance >= tolerance) {
+            return {perimeter};
+        }
+
+        Real64 min_weight = std::numeric_limits<Real64>::max();
+        int poppedVertexPos = -1;
+        int keptVertexPos = -1;
+
+        for (size_t index = 0; index < nSides; ++index) {
+            size_t const prevIndex = (index == 0) ? nSides - 1 : index - 1;
+            Real64 &distanceThisToNext = distances[index];
+            Real64 &distanceThisToPrev = distances[prevIndex];
+            if ((distanceThisToNext >= tolerance) && (distanceThisToPrev >= tolerance)) {
+                continue;
+            }
+            Real64 const weight = distanceThisToNext + distanceThisToPrev;
+            if (weight < min_weight) {
+                min_weight = weight;
+                poppedVertexPos = static_cast<int>(index);
+                if (distanceThisToPrev < distanceThisToNext) {
+                    keptVertexPos = prevIndex;
+                } else {
+                    keptVertexPos = static_cast<int>((index == nSides - 1) ? 0 : index + 1);
+                }
+            }
+        }
+
+        return {perimeter, poppedVertexPos, keptVertexPos};
```

Comment:
Pass two: figure out the vertex that is coincident with its previous and/or next vertex and that minimizes the `distanceThisToNext + distanceThisToPrev`.
We also do return the keptVertexPos (which can be the previous or the next), so we can print the displayExtraWarning correctly.

Reply:
Good approach!